src/game.py

```diff
@@ -0,0 +1,207 @@
+
+        self.__current_player: Optional[Player] = None
+        self.__waiting_players: list[Player] = []
+        self.__players_in_game: dict[int, Player] = {}
+        self.__player_wins: dict[int, int] = {}
+
+        self.__current_player_idx: Optional[int] = None
+        self.__max_players: int = max_players
+        self.__game_players: int = 0
+        self.__all_players: int = 0
+
+        self.__turn_played_sem: Semaphore = Semaphore(0)
+        self.__clock: Clock = Clock()
+
+    def add_player(self, name: str, password: str = None, is_observer: bool = None) -> None:
+        if self.__game_players >= self.__max_players:
+            is_observer = True
+
+        if not is_observer:
+            self.__game_players += 1
+
+        self.__all_players += 1
+
+        player: Player
+        player_type: str
+
+        if is_observer:
+            player_type = "observer"
+        else:
+            player_type = "bot_player"
+
+        player = PlayerFactory.create_player(player_type, name, self.__turn_played_sem, self.__current_player_idx,
+                                             self.__game_players - 1, self.running, password, is_observer)
+
+        self.__waiting_players.append(player)
+
+    def run(self) -> None:
+        self.__start_game()
+
+        while self.running:
+
+            if not self.__round_started:
+                self.__update_round()
+
+            self.__update_turn()
+
+            for event in pygame.event.get():
+                if event.type == pygame.QUIT:
+                    self.running = False
+
+            self.map.draw_map(self.current_turn, self.num_turns, self.current_round, self.num_rounds)
+
+            self.__clock.tick(60)
+
+            for player in self.__players_in_game.values():
+                player.next_turn_sem.release()
+
+            for _ in range(self.__all_players):
+                self.__turn_played_sem.acquire()
+
+        self.__end_game()
+
+    def __start_game(self) -> None:
+        self.running = True
+        self.__connect_players()
+
+        game_state: dict = self.__current_client.game_state()
+
+        self.__max_players = game_state["num_players"]
+        self.num_turns = game_state["num_turns"]
+        self.num_rounds = game_state["num_rounds"]
+
+        for idx in game_state["player_result_points"].keys():
+            self.__player_wins[int(idx)] = 0
+
+    def __update_round(self) -> None:
+        self.__round_started = True
+
+        game_map = self.__current_client.map()
+        game_state = self.__current_client.game_state()
+
+        self.current_round = game_state["current_round"]
+        for player in self.__players_in_game.values():
+            player.round_reset()
+
+        self.map = Map(game_map, game_state, self.__players_in_game)
+
+        for player in self.__players_in_game.values():
+            player.round_update(self.map)
+
+    def __update_turn(self) -> None:
+        game_state = self.__current_client.game_state()
+
```

Comment:
but if there are not only your bots in the game, there was your turn and the turn has not yet been updated, will you make a turn a second time? maybe make a check that if the turn has already been processed, then skip